### ISSUE TITLE
Xsfrm Designer generates invalid code, when "Thread missing Types as Usual (/vo15)" is deactivated in the project (x# 2.9.1.1)

### ISSUE THREAD

**Issue by VolkmarR** [X-Sharp/XSharpPublic#913]
I created a converter that deactivates the flag "Thread missing Types as Usual (/vo15)" in all our projects. The conversion works as expected, but the Xsfrm Designer causes a problem.

My converter adds the "clipper" keyword to all the methods and constructors, that don't have specified types for the parameters. 

The Constructor `constructor(oParent, uExtra)` of a dialog window is converted to `constructor(oParent, uExtra) clipper` (because otherwise I'll get a compiler error, that types are missing).

The problem is, that when I modify the form using the designer and save it, the .prg file is generated incorrectly.  The generator doesn't add the required "clipper" keyword, which results in an error.

I attached a sample project, so you can reproduce the problem:

[TestClipperConstructorDesigner.zip](https://github.com/X-Sharp/XSharpPublic/files/7875122/TestClipperConstructorDesigner.zip)

* Open the "Help About.HELPABOUT.xsfrm" file
* Save the Designer
* The "Help About.prg" get's generated without the clipper keyword
* When you build the project, you get the "Type expected" errors
* You have to add "clipper" to the constructor to be able to compile the project again

The Xsmnu Designer has the same problem.

Can this bug be fixed with the upcoming 2.10 release?

**Comment by RobertvanderHulst** [comment]
Volkmar,
We'll see if we can include this.
I do not think this is a big change, so with a little bit of luck this should work.

**Comment by VolkmarR** [comment]
Thanks.

**Comment by cpyrgas** [comment]
That wasn't really a bug (worked as in VO), but now CLIPPER calling convention clause should be added back, if it existed previously in the constructor (for all editors). Please note that it simply checks for the string "CLIPPER" after the closing paren in the constructor line, that should be enough I think.

**Comment by VolkmarR** [comment]
It might be better to check the flag in the project file. Otherwise it could be, that you have to add the clipper keyword everytime you add a new form. We use mainly our own form templates, so we can work around this issue, but it could be an issue, if someone uses one of the standard templates.

**Comment by cpyrgas** [comment]
CLIPPER is not being added always now, the editor checks if the constructor in the existing code (in the .prg file) already has the CLIPPER clause, and if it has it, only then it adds it back when it re-writes the constructor.

**Comment by VolkmarR** [comment]
Can you try to add a new form to the TestClipperConstructorDesigner project? I'm not sure, but after that, you'll probably get the missing type compile error, because I assume, that the clipper keyword is missing in the newly created prg file.

**Comment by cpyrgas** [comment]
Yes indeed, you will need to add the CLIPPER clause manually one time and then you will not need to do it again.

**Comment by VolkmarR** [comment]
A simple fix for that would be to add the clipper keyword to the standard templates. Since the constructors have to be Clipper anyway, it should not break anything, when the keyword is always present.

**Comment by cpyrgas** [comment]
Not sure, maybe Robert can introduce some conditional pieces of code in the VS templates, that can be controlled through some option somewhere in the product..

**Comment by hpetriffer** [comment]
Could it be that we just need to change cavowed.inf to make it work correctly for new windows?

**Comment by cpyrgas** [comment]
Good idea, we can introduce a new option to be defined in there. Something like an extra option at the end:

[Options]
GenerateClipperConstructors = 1

Does it sound good? You would just need to make sure this is included in all cavowed.inf files that may exist under your solution folder. Only problem is that this is used only be the window editor though, not by the menu and dbserver editors, but I assume you do not create new menus and dbservers very often...

**Comment by hpetriffer** [comment]
I have made an test and it seems that if I change
```InitMethod=dtaWindow (oWindow, iCtlID, oServer)```
to 
```InitMethod=dtaWindow (oWindow, iCtlID, oServer) clipper```
it seems to work also. Are there cases where this is not enough?

**Comment by cpyrgas** [comment]
Ah, right, this works, too! But of course that's again only for the window editor, not for the other ones. So I think the current solution should work for you, for the windows you just need to modify the cavowed.inf file accordingly, while for the menu editor etc, if you create a new menu, then you will need to apply the CLIPPER clause manually. But subsequent edits of the menu will retain this in the code. Is that correct?

**Comment by VolkmarR** [comment]
Yes, that should work. It is very rare, that we add a new menu.

**Comment by cpyrgas** [comment]
OK, so we already have a solution then! For windows it already works for you now by adjusting cavowed.inf, while for the menu entities in the new build the CLIPPER clause will be respected if already found in the code.

**Comment by cpyrgas** [comment]
I have made sure now that CLIPPER does not get added twice if you have already specified it in cavowed.inf. If you'd like, you can already test the update, by putting this dll to C:\Program Files (x86)\XSharp\Extension\Project, then open a command prompt in ADMIN mode, navigate to C:\Program Files (x86)\XSharp\Uninst and run deployvs1.cmd for VS2017, deployvs2.cmd for VS2019:

[XSharp.CodeGenerator.zip](https://github.com/X-Sharp/XSharpPublic/files/7884596/XSharp.CodeGenerator.zip)

**Comment by hpetriffer** [comment]
@cpyrgas I have tested it with the last version which I got yesterday from Robert. In this version it seems to work.

**Comment by cpyrgas** [comment]
OK, thanks for feedback!
